js/noms/src/spec.js

```diff
@@ -0,0 +1,315 @@
+   */
+  static forDataset(spec: string, opts: ?SpecOptions): Spec {
+    const [databaseSpec, datasetName] = splitDatabaseSpec(spec);
+    if (!datasetRe.test(datasetName)) {
+      throw new SyntaxError(`Invalid dataset ${datasetName}, must match ${datasetRe.source}`);
+    }
+
+    const specObj = new Spec(spec, databaseSpec, opts);
+    specObj._datasetName = datasetName;
+    return specObj;
+  }
+
+  /**
+   * Creates a `Spec` from a spec to a value path.
+   */
+  static forPath(spec: string, opts: ?SpecOptions): Spec {
+    const [databaseSpec, pathStr] = splitDatabaseSpec(spec);
+    const path = AbsolutePath.parse(pathStr);
+
+    const specObj = new Spec(spec, databaseSpec, opts);
+    specObj._path = path;
+    return specObj;
+  }
+
+  /**
+   * Returns the spec string.
+   */
+  spec(): string {
+    return this._spec;
+  }
+
+  /**
+   * Returns the database protocol, for example `"mem"` or `"http"`.
+   */
+  protocol(): string {
+    return this._protocol;
+  }
+
+  /**
+   * Returns the database name, e.g. `"demo.noms.io"` or `"localhost:8000"`.
+   */
+  databaseName(): string {
+    return this._databaseName;
+  }
+
+  /**
+   * Returns the `Database`.
+   */
+  database(): Database {
+    if (!this._database) {
+      this._database = this._createDatabase();
+    }
+    return this._database;
+  }
+
+  /**
+   * Returns the dataset name, for example `"sf-crime"`.
+   *
+   * Throws a `TypeError` if this spec does not have an associated value.
+   */
+  datasetName(): string {
+    const name = this._datasetName;
+    if (name === undefined || name === null) {
+      throw new TypeError(`${this._spec} is not a dataset spec`);
+    }
+    return name;
+  }
+
+  /**
+   * Returns the path with this spec's `Database`. Throws a `TypeError` if this
+   * spec does not have an associated path.
+   */
+  path(): AbsolutePath {
+    if (!this._path) {
+      throw new TypeError(`${this._spec} is not a path spec`);
+    }
+    return this._path;
+  }
+
+  /**
+   * Returns the `Dataset` from the spec.
+   *
+   * Throws a `TypeError` if this spec does not have a associated dataset.
+   *
+   * Make sure to call `close()`.
+   */
+  dataset(): Dataset {
+    return this.database().getDataset(this.datasetName());
+  }
+
+  /**
+   * Resolves this to a `Value`, or `null` if the value isn't found.
+   *
+   * Throws a `TypeError` if this spec does not have an associated value.
+   *
+   * Make sure to call `close()`.
+   */
+  value(): Promise<Value | null> {
+    return this.path().resolve(this.database());
+  }
+
+  /**
+   * Returns a `Spec` in which the dataset component, if any, has been replaced
+   * with the hash of the HEAD of that dataset. This "pins" the path to the
+   * state of the database at the current moment in time.
+   *
+   * Returns this if the PathSpec is already "pinned".
+   *
+   * Throws a `TypeError` if this spec does not have an associated dataset.
+   */
+  async pin(): Promise<?Spec> {
```

Comment:
The only real difference between spec.js and specs.js is the implementation of `pin()`. It now supports also pinning Datasets, as opposed to an AbsolutePath from a dataset. This didn't make sense before because `pin()` was on `PathSpec`.

Tests have been added accordingly.